File: go/2024/day19/cmd/main.go
package main

import (
	"fmt"
	"helpers"
	"strings"
)

const Towels = "r, wr, b, g, bwu, rb, gb, br"

func completePattern(pattern string) bool {
	if pattern == "" {
		return true
	}

	done := false
	for _, towel := range strings.Split(Towels, ", ") {
		if done {
			return done
		}

		if strings.HasPrefix(pattern, towel) {
			done = completePattern(pattern[len(towel):])
		}
	}

	return false
}

func main() {
	input := helpers.ReadFile("../input.txt")
	part1 := 0

	for _, pattern := range input[2:] {
		if completePattern(pattern) {
			part1 += 1
		}
	}

	fmt.Printf("Part 1: %d\n", part1)
}

File: zig/2024/day19/main.zig
const std = @import("std");
const print = std.debug.print;
const assert = std.debug.assert;
const file = @embedFile("input.txt");

const Towels = "r, wr, b, g, bwu, rb, gb, br";

const test_input =
    \\brwrr
    \\bggr
    \\gbbr
    \\rrbgbr
    \\ubwu
    \\bwurrg
    \\brgr
    \\bbrgwb
;

fn completePattern(pattern: []const u8) bool {
    if (pattern.len == 0) {
        return true;
    }

    var done: bool = false;
    var towels = std.mem.tokenizeAny(u8, Towels, ", ");
    while (towels.next()) |towel| {
        if (done) {
            return done;
        }

        if (std.mem.startsWith(u8, pattern, towel)) {
            done = completePattern(pattern[towel.len..]);
        }
    }

    return false;
}

pub fn main() void {
    var part1: usize = 0;

    var tok = std.mem.tokenizeAny(u8, test_input, "\n");
    while (tok.next()) |pattern| {
        if (completePattern(pattern)) {
            part1 += 1;
        }
    }

    print("Part 1: {d}\n", .{part1});
}
